container/lru/map.go

```diff
@@ -0,0 +1,524 @@
+// Copyright (c) 2024 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+// Package lru provides generic type and concurrent safe LRU data structures
+// with near O(1) perf and optional time-based expiration support.
+package lru
+
+import (
+	"sync"
+	"time"
+)
+
+// expireScanInterval is the minimum interval between expiration scans.
+const expireScanInterval = 30 * time.Second
+
+// element is an element in the linked list used to house the KV pairs and
+// associated expiration data.
+type element[K comparable, V any] struct {
+	key          K
+	value        V
+	expiresAfter int64
+	prev, next   *element[K, V]
+}
+
+// Map provides a concurrency safe least recently used (LRU) map with nearly
+// O(1) lookups, inserts, and deletions.  The map is limited to a maximum number
+// of items with eviction of the least recently used entry when the limit is
+// exceeded.
+//
+// It also supports optional item expiration after a configurable time to live
+// (TTL) with periodic lazy removal.  [NewMap] provides no item expiration by
+// default while [NewMapWithDefaultTTL] provides a configurable TTL that is
+// applied to all items by default.  In both cases, callers may override the
+// default TTL of individual items via [PutWithTTL].
+//
+// Expiration TTLs are relative to the time an item is added or updated via
+// [Put] or [PutWithTTL].  Accessing items does not extend the TTL.
+//
+// An efficient lazy removal scheme is used such that expired items are
+// periodically removed when items are added or updated via [Put] or
+// [PutWithTTL].  In other words, expired items may physically remain in the map
+// until the next expiration scan is triggered, however, they will no longer
+// publicly appear as members of the map.  This approach allows for efficient
+// amortized removal of expired items without the need for additional background
+// tasks or timers.
+//
+// Callers may optionally use [EvictExpiredNow] to immediately remove all items
+// that are marked expired without waiting for the next expiration scan in cases
+// where more deterministic expiration is required.
+//
+// [NewMap] or [NewMapWithDefaultTTL] must be used to create a usable map since
+// the zero value of this struct is not valid.
+type Map[K comparable, V any] struct {
+	// These fields do not require a mutex since they are set at initialization
+	// time and never modified after.
+	nowFn func() time.Time
+	limit uint32
+	ttl   time.Duration
+
+	mtx   sync.Mutex
```

Comment:
If this were updated to `RWMutex`, funcs like `Peek` and `Exist` which dont modify the linked list could `RLock` instead of `Lock`, enabling them to be run concurrently.

Reply:
I tend to go back and forth a bit on the usage of `RWMutex`.  The tricky part is that it has more overhead than `Mutex` and because it's often the case that accesses aren't actually highly concurrent, `Mutex`, perhaps counterintuitively, ends up being slightly faster overall.

However, since this aims to be an entirely generic LRU where we don't know for sure if the caller will actually be highly concurrent of not, I think you're right that `RWMutex` makes more sense here.  It allows the possibility for getting much higher throughput in highly concurrent scenarios at only a very minor cost to the less concurrent case.

Good call!  Updated.